### PR TITLE
Refactor emit_texture_op() function.

### PR DESCRIPTION
**Pull request by billhollings**
Refactor emit_texture_op() function to simplify subclass overrides.
CompilerMSL refactored to simplify handling of vertex attributes.

### CODE REVIEW

**Comment by HansKristian-ARM** [comment]
The refactor looks good, but getting a lot of warnings related to unused parameters on GCC 6.2.1.

```
spirv_msl.cpp: In member function ‘virtual std::__cxx11::string spirv_cross::CompilerMSL::to_function_name(uint32_t, const spirv_cross::SPIRType&, bool, bool, bool, bool, bool, bool, bool, bool)’:
spirv_msl.cpp:702:68: warning: unused parameter ‘imgtype’ [-Wunused-parameter]
 string CompilerMSL::to_function_name(uint32_t img, const SPIRType &imgtype, bool is_fetch, bool is_gather, bool is_proj,
                                                                    ^~~~~~~
spirv_msl.cpp:702:113: warning: unused parameter ‘is_proj’ [-Wunused-parameter]
 32_t img, const SPIRType &imgtype, bool is_fetch, bool is_gather, bool is_proj,
                                                                        ^~~~~~~
spirv_msl.cpp:703:43: warning: unused parameter ‘has_array_offsets’ [-Wunused-parameter]
                                      bool has_array_offsets, bool has_offset, bool has_grad, bool has_lod,
                                           ^~~~~~~~~~~~~~~~~
spirv_msl.cpp:703:67: warning: unused parameter ‘has_offset’ [-Wunused-parameter]
                                      bool has_array_offsets, bool has_offset, bool has_grad, bool has_lod,
                                                                   ^~~~~~~~~~
spirv_msl.cpp:703:84: warning: unused parameter ‘has_grad’ [-Wunused-parameter]
                       bool has_array_offsets, bool has_offset, bool has_grad, bool has_lod,
                                                                     ^~~~~~~~
spirv_msl.cpp:703:99: warning: unused parameter ‘has_lod’ [-Wunused-parameter]
           bool has_array_offsets, bool has_offset, bool has_grad, bool has_lod,
                                                                        ^~~~~~~
spirv_msl.cpp: In member function ‘virtual std::__cxx11::string spirv_cross::CompilerMSL::to_function_args(uint32_t, const spirv_cross::SPIRType&, bool, bool, bool, uint32_t, uint32_t, uint32_t, uint32_t, uint32_t, uint32_t, uint32_t, uint32_t, uint32_t, uint32_t, uint32_t, bool*)’:
spirv_msl.cpp:724:97: warning: unused parameter ‘is_gather’ [-Wunused-parameter]
 ion_args(uint32_t img, const SPIRType &imgtype, bool is_fetch, bool is_gather, bool is_proj,
                                                                     ^~~~~~~~~
spirv_msl.cpp:725:63: warning: unused parameter ‘coord_components’ [-Wunused-parameter]
                                      uint32_t coord, uint32_t coord_components, uint32_t dref, uint32_t grad_x,
                                                               ^~~~~~~~~~~~~~~~
spirv_msl.cpp:727:62: warning: unused parameter ‘sample’ [-Wunused-parameter]
                                      uint32_t comp, uint32_t sample, bool *p_forward)
                                                              ^~~~~~
g++ -c -o spirv_cpp.o spirv_cpp.cpp -std=c++11 -Wall -Wextra -Wshadow -D__STDC_LIMIT_MACROS -O2 -DNDEBUG -MMD
g++ -c -o spirv_cfg.o spirv_cfg.cpp -std=c++11 -Wall -Wextra -Wshadow -D__STDC_LIMIT_MACROS -O2 -DNDEBUG -MMD
g++ -c -o spirv_glsl.o spirv_glsl.cpp -std=c++11 -Wall -Wextra -Wshadow -D__STDC_LIMIT_MACROS -O2 -DNDEBUG -MMD
spirv_glsl.cpp: In member function ‘virtual std::__cxx11::string spirv_cross::CompilerGLSL::to_function_name(uint32_t, const spirv_cross::SPIRType&, bool, bool, bool, bool, bool, bool, bool, bool)’:
spirv_glsl.cpp:2561:48: warning: unused parameter ‘img’ [-Wunused-parameter]
 string CompilerGLSL::to_function_name(uint32_t img, const SPIRType &imgtype, bool is_fetch, bool is_gather,
                                                ^~~
spirv_glsl.cpp:2563:58: warning: unused parameter ‘has_dref’ [-Wunused-parameter]
                                       bool has_lod, bool has_dref)
                                                          ^~~~~~~~
spirv_glsl.cpp: In member function ‘virtual std::__cxx11::string spirv_cross::CompilerGLSL::to_function_args(uint32_t, const spirv_cross::SPIRType&, bool, bool, bool, uint32_t, uint32_t, uint32_t, uint32_t, uint32_t, uint32_t, uint32_t, uint32_t, uint32_t, uint32_t, uint32_t, bool*)’:
spirv_glsl.cpp:2592:69: warning: unused parameter ‘imgtype’ [-Wunused-parameter]
 string CompilerGLSL::to_function_args(uint32_t img, const SPIRType &imgtype, bool is_fetch, bool is_gather,
                                                                     ^~~~~~~
spirv_glsl.cpp:2592:83: warning: unused parameter ‘is_fetch’ [-Wunused-parameter]
 string CompilerGLSL::to_function_args(uint32_t img, const SPIRType &imgtype, bool is_fetch, bool is_gather,
                                                                                   ^~~~~~~~
spirv_glsl.cpp:2592:98: warning: unused parameter ‘is_gather’ [-Wunused-parameter]
 string CompilerGLSL::to_function_args(uint32_t img, const SPIRType &imgtype, bool is_fetch, bool is_gather,
                                                                                                  ^~~~~~~~~
spirv_glsl.cpp:2593:44: warning: unused parameter ‘is_proj’ [-Wunused-parameter]
                                       bool is_proj, uint32_t coord, uint32_t coord_components, uint32_t dref,
                                            ^~~~~~~
```
Just drop the parameter name to fix that.

File: spirv_glsl.hpp [hunk 1]
```diff
@@ -168,6 +168,7 @@ class CompilerGLSL : public Compiler
 	                                     uint32_t grad_x, uint32_t grad_y, uint32_t lod, uint32_t coffset,
 	                                     uint32_t offset, uint32_t bias, uint32_t comp, uint32_t sample,
 	                                     bool *p_forward);
+    virtual std::string clean_func_name(std::string func_name);
```
Comment (HansKristian-ARM) [inline spirv_glsl.hpp]: Formatting?

File: spirv_msl.hpp [hunk 2]
```diff
@@ -148,6 +149,7 @@ class CompilerMSL : public CompilerGLSL
 	std::string to_component_argument(uint32_t id);
 
 	MSLConfiguration msl_config;
+    std::unordered_map<std::string,std::string> func_name_overrides;
```
Comment (HansKristian-ARM) [inline spirv_msl.hpp]: space between the two std::strings.

**Comment by billhollings** [comment]
Sorry...my bad. I forgot to run the formatter before committing. Done now.

**Pull request merged**